aptos-move/framework/src/natives/string_utils.rs

```diff
@@ -0,0 +1,322 @@
+            out.push('}');
+        },
+        MoveTypeLayout::Struct(MoveStructLayout::Runtime(fields)) => {
+            let strct = val.value_as::<Struct>()?;
+            out.push('{');
+            format_vector(
+                context,
+                base_gas,
+                fields.iter().zip(strct.unpack()?).collect(),
+                out,
+            )?;
+            out.push('}');
+        },
+    };
+    Ok(())
+}
+
+pub fn native_format(
+    gas_params: &GasParameters,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert!(ty_args.len() == 1);
+    let ty = context
+        .deref()
+        .type_to_fully_annotated_layout(&ty_args[0])?
+        .unwrap();
+    let x = safely_pop_arg!(arguments, Reference);
+    let v = x.read_ref().map_err(SafeNativeError::InvariantViolation)?;
+    let mut out = String::new();
+    native_format_impl(context, gas_params.base, &ty, v, &mut out)?;
+    let move_str = Value::struct_(Struct::pack(vec![Value::vector_u8(out.into_bytes())]));
+    Ok(smallvec![move_str])
+}
+
+pub fn native_format_list(
+    gas_params: &GasParameters,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert!(ty_args.len() == 1);
+    let mut list_ty = &ty_args[0];
+
+    let arg_mismatch = 1;
+    let invalid_fmt = 2;
+
+    let val = safely_pop_arg!(arguments, Reference);
+    let mut val = val
+        .read_ref()
+        .map_err(SafeNativeError::InvariantViolation)?;
+
+    let fmt = safely_pop_arg!(arguments, Reference);
+    let fmt = fmt
+        .read_ref()
+        .map_err(SafeNativeError::InvariantViolation)?
+        .value_as::<Vec<u8>>()?;
+    let fmt = std::str::from_utf8(&fmt).map_err(|_| SafeNativeError::Abort {
+        abort_code: invalid_fmt,
+    })?;
+
+    context.charge(gas_params.per_byte * GasQuantity::from(fmt.len() as u64))?;
+
+    let match_list_ty = |context: &mut SafeNativeContext, list_ty, name| {
+        if let TypeTag::Struct(struct_tag) = context
+            .type_to_type_tag(list_ty)
+            .map_err(SafeNativeError::InvariantViolation)?
+        {
+            if !(struct_tag.address == AccountAddress::ONE
+                && struct_tag.module.as_str() == "string_utils"
+                && struct_tag.name.as_str() == name)
+            {
+                return Err(SafeNativeError::Abort {
+                    abort_code: arg_mismatch,
+                });
+            }
+            Ok(())
+        } else {
+            Err(SafeNativeError::Abort {
+                abort_code: arg_mismatch,
+            })
+        }
+    };
+
+    let mut out = String::new();
+    let mut in_braces = false;
+    let mut in_escape = false;
+    for c in fmt.chars() {
+        if !in_escape && c == '\\' {
```

Comment:
Nit: this can be nicer done with a match, perhaps.


```
match (in_escape, c) {
  (false, '\\') => { in_escape = true; continue }
  (false, '{') => ..
```

Reply:
Rust uses {{ and }} for escaping brackets so I'm using the same now